xfrd.c

```diff
@@ -2739,6 +2741,70 @@ xfrd_handle_taskresult(xfrd_state_type* xfrd, struct task_list_d* task)
 		xfrd_process_zonestat_inc_task(xfrd, task);
 		break;
 #endif
+	case task_add_catzone: {
+		char* zname = (char*)task->zname;
+		char* pattern = zname + strlen(zname) + 1;
+		char* catname = pattern + strlen(pattern) + 1;
+		char* member_id = catname + strlen(catname) + 1;
+
+		struct zone_options* zopt = NULL;
+		pattern_options_type* patopt;
+
+		log_msg(LOG_INFO, "now adding catzone %s %s %s %s", zname, pattern, catname, member_id);
+
+		patopt = pattern_options_find(xfrd->nsd->options, pattern);
+		if (!patopt) {
+			patopt = pattern_options_create(xfrd->region);
+			patopt->pname = pattern;
+			pattern_options_add_modify(xfrd->nsd->options, patopt);
+		} 
+		zopt = zone_options_find(xfrd->nsd->options, dname_parse(xfrd->region, zname));
+		if (!zopt) {
+			zopt = zone_list_zone_insert(xfrd->nsd->options, zname, pattern, 0, 0);
+		} else {
+			log_msg(LOG_INFO, "catzone already added %s", zname);
+			return;
+		}
+		task_new_add_catzone(xfrd->nsd->task[xfrd->nsd->mytask],
+			xfrd->last_task, zname, pattern, catname, member_id,
+			getzonestatid(xfrd->nsd->options, zopt));
+		xfrd_set_reload_now(xfrd);
+		/* add to xfrd - notify (for master and slaves) */
+		init_notify_send(xfrd->notify_zones, xfrd->region, zopt);
+		/* add to xfrd - slave */
+		xfrd_init_slave_zone(xfrd, zopt);
```

Comment:
It may be nice to guard this call to `xfrd_init_slave_zone` with an if on `zone_is_slave(zopt)` , so that a catalog zone could add primary zones. If the zone is primary depends on the config for it and this is determined by the config pattern that is used, and since that can be set, it seems possible.

Reply:
This item seems to be a duplicate comment, that was generated as a copy from another. I'll close it.